urllib3/transport_security.py

```diff
@@ -0,0 +1,67 @@
+import logging
+from six.moves.http_cookiejar import split_header_words
+from . import util
+
+log = logging.getLogger(__name__)
+
+
+class TransportSecurityManager(object):
+    """
+    Coordinates HSTS, HPKP, and certificate fingerprint matching
+    on outgoing requests and incoming responses.
+    """
+    def __init__(self, transport_security_store=None):
+        self._tss = transport_security_store or TransportSecurityStore()
+        log.debug("tss init")
+
+    def validate_new_connection(self, conn, scheme):
+        """
+        Enforce HSTS.
+        """
+        log.debug("hsts validate: %s %s %s", scheme, conn.host, conn.port)
+
+    def validate_established_connection(self, conn):
+        """
+        Enforce HPKP or a custom certificate fingerprint.
+        """
+        if util.IS_PYOPENSSL:
+            log.debug("hpkp validate: %s", conn.connection.certs)
+        else:
+            log.debug("PyOpenSSL not available, hpkp validation disabled")
```

Comment:
Fundamentally, HSTS is concerned with examining the request before the connection is made (to prevent any unsecured content from going out over the wire), and HPKP is concerned with examining the completed TLS connection before request data is sent. So I think both these methods are necessary. We could make them into one method, but then it would have to be called from two places and adjust its behavior based on whether the connection is live.